lib/interface/cli/commands/offline-logs/offload-to-collection.cmd.js

```diff
@@ -1,96 +1,278 @@
 const { MongoClient, ObjectId } = require("mongodb");
-const moment = require('moment')
-const Command = require('../../Command');
-const cmd = require('./base.cmd');
-const { objectIdFromDate, defaultCollections } = require('./utils')
-
-const offloadToCollection = async function(sourceDBObj, collection, targetDB, cutoffDate) {
-  const sourceCollectionObj = sourceDBObj.collection(collection)
-  const targetCollection = `archive-${collection}`
-  
-  const cutoffDateObj = moment(cutoffDate)
-                        .add(1, 'days')
-                        .startOf("day")
-
-  if(!cutoffDateObj.isValid()){
-    throw new Error('please enter a valid date in ISO 8601 format')
+const moment = require("moment");
+const semver = require("semver");
+const Command = require("../../Command");
+const cmd = require("./base.cmd");
+const {
+  objectIdFromDate,
+  defaultCollections,
+  collectionExists,
+  findMinLog,
+} = require("./utils");
+
+const offloadToCollection = async function (
+  sourceDBObj,
+  collection,
+  targetDB,
+  cutoffDate
+) {
+  const sourceCollectionObj = sourceDBObj.collection(collection);
+  const targetCollection = `archive-${collection}`;
+
+  const cutoffDateObj = moment(cutoffDate).add(1, "days").startOf("day");
+
+  if (!cutoffDateObj.isValid()) {
+    throw new Error("please enter a valid date in ISO 8601 format");
+  }
+
+  const cutoffDateId = objectIdFromDate(cutoffDateObj.toDate());
+
+  const mongoInfo = await sourceDBObj.admin().serverInfo();
+  const mongoServerVersion = await mongoInfo.version;
+  const mergeReleaseVersion = "4.2.0";
+
+  if (semver.lte(mergeReleaseVersion, mongoServerVersion)) {
+    await archiveWithMerge(
+      sourceCollectionObj,
+      cutoffDateId,
+      targetDB,
+      targetCollection,
+      collection
+    );
+  } else {
+    if (await collectionExists(sourceDBObj, targetCollection)) {
+      await archiveInChunks(
+        sourceDBObj,
+        sourceCollectionObj,
+        targetCollection,
+        cutoffDateId,
+        collection,
+        cutoffDateObj
+      );
+    } else {
+      await archiveWithOut(
+        sourceCollectionObj,
+        cutoffDateId,
+        targetCollection,
+        collection
+      );
+    }
+  }
+};
+///////////////////////////////////////////////////////////////////////////////////
+const archiveWithMerge = async function (
+  sourceCollectionObj,
+  cutoffDateId,
+  targetDB,
+  targetCollection,
+  collection
+) {
+  const minLog = await findMinLog(
+    sourceCollectionObj,
+    cutoffDateId,
+    collection
+  );
+
+  if (!minLog) {
+    console.info(
+      `No logs to archive in Collection ${collection} from the given date.`
+    );
+    return;
+  }
```

Comment:
you can extract this bit to the parent function, and only have this code in a single place.

Reply:
done